### PR TITLE
[AMD-ZENDNN] Modifications to support eager op rewrite for AMD CPU Inference

### PR DESCRIPTION
**Pull request by sanadani**
This pull request introduces support to re-write eager ops with ZenOps.
All these code changes are under AMD_ZENDNN flag which is enabled for linux builds only.

Authors:
Aakar Dwivedi ( aakar.dwivedi@amd.com )
Avinash-Chandra Pandey ( avinash-chandra.pandey@amd.com )
Chandra Kumar Ramasamy ( chandrakumar.ramasamy@amd.com )
Savan Anadani ( savan.anadani@amd.com )

Signed-off-by: sanadani <savan.anadani@amd.com>

### CODE REVIEW

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  return (*new_zen_op)->SetDeviceName(device_name.c_str());
+}
+
+Status ZenEagerOpRewrite::CreateGenericZenOp(
+    EagerOperation *orig_op, std::unique_ptr<EagerOperation> *zen_op,
+    string zen_op_name) {
+  VLOG(1) << " TF-EAGER-REWRITE Info: OriginalOp= " << orig_op->Name()
+          << " ZenOp=" << zen_op_name;
+
+  TF_CHECK_OK(SetupNewOp(orig_op, zen_op_name, zen_op));
```
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:152]: Right now `SetupNewOp` is only used once here. `CreateGenericZenOp` also doesn't do much apart from calling `SetupNewOp`. Do you plan to have another Zen-op creation function that will call `SetupNewOp` in the future? If not, let's just merge `SetupNewOp` and `CreateGenericZenOp` into one function.

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  return OkStatus();
+}
+
+bool ZenEagerOpRewrite::ShouldRewriteOp(EagerOperation *op) {
+  // Don't rewrite the op if TF-ZenDNN use is disabled at runtime.
+  if (!IsZenDnnEnabled()) {
+    return false;
+  }
```
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:160]: Let's do this check at the `ZenEagerOpRewrite::Run` level instead.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the PR! I have minor comments.

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  return OkStatus();
+}
+
+Status ZenEagerOpRewrite::SetupNewOp(
+    EagerOperation *orig_op, string zen_op_name,
+    std::unique_ptr<EagerOperation> *new_zen_op) {
+  bool is_remote = false;
+  new_zen_op->reset(new tensorflow::EagerOperation(&orig_op->EagerContext()));
+  TF_RETURN_IF_ERROR(new_zen_op->get()->Reset(zen_op_name.c_str(), nullptr,
+                                              is_remote, nullptr));
```
[the remark this answers is not in the document]
Comment (sanadani) [reply tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc]: Fixed. Thank you!

**Review by penpornk — APPROVED** [review]
Thank you very much for the quick changes!

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  if (it != zen_eager_ops_.end()) {
+    // Eager op found, verify that a kernel exists for Zen op and rewrite is
+    // possible.
+    if (zen_op_registry::IsZenOpKernelRegistered(
+            zen_op_registry::GetZenOpName(op->Name()), data_type) &&
+        it->second.rewrite_rule(op)) {
+      return true;
+    }
+  }
+  return false;
```
[the remark this answers is not in the document]
Comment (sanadani) [reply tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc]: Thank you for noticing that and letting us know. Accepted the change.

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+void ZenEagerOpRewrite::InsertZenEagerOps(ZenEagerOp op) {
+  zen_eager_ops_.insert(std::make_pair(op.op_name, op));
+}
+
+Status ZenEagerOpRewrite::Run(
+    EagerOperation *orig_op,
+    std::unique_ptr<tensorflow::EagerOperation> *out_op) {
+  if (ShouldRewriteOp(orig_op)) {
```
[the remark this answers is not in the document]
Comment (sanadani) [reply tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc]: Accepted the change. Thank you!

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+  int num_inputs = orig_op->Inputs().size();
+  // Add all inputs to the new op.
+  for (int i = 0; i < num_inputs; ++i) {
+    TF_RETURN_IF_ERROR((*new_zen_op)->AddInput(orig_op->Inputs()[i]));
+  }
+
+  // Copy all attributes to the new op.
+  const NodeDef &kOrigNodeDef = orig_op->MutableAttrs()->BuildNodeDef();
+
```
[the remark this answers is not in the document]
Comment (sanadani) [reply tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc]: Fixed. Thank you!

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc (hunk 2 above)
Comment (sanadani) [reply tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc]: Accepted the change. Thank you!

File: tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc (hunk 1 above)
Comment (sanadani) [reply tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc]: Accepted the change. Thank you!

**Comment by penpornk** [comment]
Got [AMD ROCm test failure](http://ml-ci.amd.com:21096/blue/organizations/jenkins/tensorflow%2Fgithub-prs-upstream-master%2FAMD-ROCm-Community-CI-Build/detail/PR-59314/3/pipeline)
```
ERROR: /workspace/tensorflow/core/common_runtime/eager/BUILD:518:11: no such target '//tensorflow/core/graph:zen_graph_util': target 'zen_graph_util' not declared in package 'tensorflow/core/graph' (did you mean 'mkl_graph_util'?) defined by /workspace/tensorflow/core/graph/BUILD and referenced by '//tensorflow/core/common_runtime/eager:execute'
```
I guess we'll have to wait for #59137 to be merged, sync this PR (if needed), then rerun all the tests.

**Comment by sanadani** [comment]
Thank you @penpornk  so much for quick review.

The PR https://github.com/tensorflow/tensorflow/pull/59137 is merged. 
@penpornk @gbaned Can you please re-trigger the CI Builds and other tests for this PR?

Thank you!

**Comment by penpornk** [comment]
@sanadani Triggering them now. :) If they still can't find `//tensorflow/core/graph:zen_graph_util`, you might have to rebase the PR. We should know in an hour.

**Comment by sanadani** [comment]
Hi @penpornk,

Thank you for triggering run quickly.

It looks like the build was successful. But the wheel file installation [failed ](https://github.com/tensorflow/tensorflow/actions/runs/3994639680/jobs/6852582775#step:5:16835) 

Can you please check? We request for any detail/direction that could help us to know more about this issue and resolve it.

**Comment by penpornk** [comment]
@sanadani Thank you for checking! [The Arm CI failure](https://github.com/tensorflow/tensorflow/actions/runs/3994639680/jobs/6852582775) is unrelated to this PR. It failed building `h5py` and not `tensorflow`. 
```
ERROR: Failed building wheel for h5py
```
The Arm CI pipeline has been failing [for the past few days](https://github.com/tensorflow/tensorflow/actions/workflows/arm-ci.yml) so we should just ignore it.

I'm waiting for the 1MacOS CPU Python3` and `Windows Bazel` CI to be done and then I'll import the PR. :)

**Pull request merged**